lib/upgrade/etcd/conversionv1v3/workloadendpoint.go

```diff
@@ -0,0 +1,271 @@
+
+	ipNats := convertIPNATs(wepValue.IPv4NAT)
+	ipNats = append(ipNats, convertIPNATs(wepValue.IPv6NAT)...)
+
+	wep := apiv3.NewWorkloadEndpoint()
+
+	wep.ObjectMeta = v1.ObjectMeta{
+		Namespace: namespace,
+		Labels:    labels,
+	}
+	wep.Spec = apiv3.WorkloadEndpointSpec{
+		Orchestrator:  convertName(wepKey.OrchestratorID),
+		Workload:      workload,
+		Node:          convertName(wepKey.Hostname),
+		Pod:           pod,
+		ContainerID:   container,
+		Endpoint:      convertName(wepKey.EndpointID),
+		IPNetworks:    ipNets,
+		IPNATs:        ipNats,
+		Profiles:      convertProfiles(wepValue.ProfileIDs),
+		InterfaceName: wepValue.Name,
+		Ports:         convertPorts(wepValue.Ports),
+	}
+
+	if wepValue.IPv4Gateway != nil {
+		wep.Spec.IPv4Gateway = wepValue.IPv4Gateway.String()
+	}
+	if wepValue.IPv6Gateway != nil {
+		wep.Spec.IPv6Gateway = wepValue.IPv6Gateway.String()
+	}
+	if wepValue.Mac != nil {
+		wep.Spec.MAC = wepValue.Mac.String()
+	}
+
+	// Figure out the new name based on WEP fields.
+	wepids := names.WorkloadEndpointIdentifiers{
+		Node:         wep.Spec.Node,
+		Orchestrator: wep.Spec.Orchestrator,
+		Endpoint:     wep.Spec.Endpoint,
+		Workload:     wep.Spec.Workload,
+		Pod:          wep.Spec.Pod,
+		ContainerID:  wep.Spec.ContainerID,
+	}
+
+	name, err := wepids.CalculateWorkloadEndpointName(false)
+	if err != nil {
+		return nil, err
+	}
+	wep.ObjectMeta.Name = name
+
+	log.Debugf("Converted: %+v\n To: %+v", kvp, wep)
+
+	return wep, nil
+}
+
+func convertMetadataToKey(m unversioned.ResourceMetadata) (model.Key, error) {
+	hm := m.(apiv1.WorkloadEndpointMetadata)
+	k := model.WorkloadEndpointKey{
+		Hostname:       hm.Node,
+		OrchestratorID: hm.Orchestrator,
+		WorkloadID:     hm.Workload,
+		EndpointID:     hm.Name,
+	}
+	return k, nil
+}
+
+// convertLabels creates a new map of labels, and updates the v1 namespace label to the v3 style.
+func convertLabels(v1Labels map[string]string) map[string]string {
+	labels := map[string]string{}
+	for k, v := range v1Labels {
+		labels[k] = v
+	}
+
+	if val, ok := labels["calico/k8s_ns"]; ok {
+		labels["projectcalico.org/namespace"] = val
+		delete(labels, "calico/k8s_ns")
+	}
+
+	return labels
+}
+
+// convertIPNetworks updates the old []net.IPNet to []string.
+func convertIPNetworks(ipNetworks []net.IPNet) []string {
+	var ipNets []string
+	for _, ipNet := range ipNetworks {
+		ipNets = append(ipNets, ipNet.String())
+	}
+
+	return ipNets
+}
+
+// convertIPNATs updates the type of IPNAT struct used.
+func convertIPNATs(v1IPNATs []model.IPNAT) []apiv3.IPNAT {
+	var ipNATs []apiv3.IPNAT
+	for _, ipNAT := range v1IPNATs {
+		ipNATs = append(ipNATs, apiv3.IPNAT{
+			InternalIP: ipNAT.IntIP.String(),
+			ExternalIP: ipNAT.ExtIP.String(),
+		})
+	}
+
+	return ipNATs
+}
+
+// convertProfiles updates the Kubernetes namespace portion from "k8s_ns" to "kns" for each profile.
+func convertProfiles(v1Profiles []string) []string {
```

Comment:
I think this function should also pass the names through `convertName`. Old profile names pass through `namespacedNameRegex` which allows the `/` character. But the new names don't allow it.